Handle importing globals in s2wasm

jgravelle-google opened the pull request:
Depends on https://reviews.llvm.org/D26875

kripken commented on src/wasm-emscripten.cpp:
```diff
@@ -86,6 +86,17 @@ std::vector<Function*> makeDynCallThunks(Module& wasm, std::vector<Name> const&
 }
 
+void addGlobalImports(Module& wasm, std::unordered_set<cashew::IString> const& globalImports) {
+  for (Name name : globalImports) {
```
how about `auto` here?

jgravelle-google replied:
Done.
Turns out `Name` didn't have an implementation of `hash` and `equal_to` whereas `cashew::IString` did, so I added those to `Name` so we could use it in sets.

kripken commented:
lgtm (but should probably wait for @dschuff for the linking-specific stuff?), and nice about the hashing.

dschuff commented on src/wasm.h:
```diff
@@ -1619,6 +1619,18 @@ template<> struct hash<wasm::Address> {
 };
-}
+
+template <> struct hash<wasm::Name> : public unary_function<wasm::Name, size_t> {
```
Could we split this into its own change? There's a bunch of cases where we use `cashew::IString` instead of `Name` in unordered sets and the like because of this issue and we can fix those along with it, but we probably don't want those in this change.

dschuff commented on src/s2wasm.h:
```diff
@@ -416,6 +415,10 @@ class S2WasmBuilder {
           abort_on("unknown directive");
         }
       // add data aliases
```
I guess this comment should be moved down

dschuff commented on src/wasm-linker.h:
```diff
@@ -86,18 +89,18 @@ class LinkerObject {
-
-  Relocation* getCurrentRelocation() {
-    return relocations.back().get();
+  void addRelocation(Relocation* relocation) {
```
Should add a comment that this takes ownership of `relocation`

dschuff commented on src/wasm-linker.h:
```diff
@@ -86,18 +89,18 @@ class LinkerObject {
     return symbolInfo.implementedFunctions.count(name) != 0;
   }
 
+  bool isGlobalImported(Name name) {
```
This is backwards compared to `isFunctionImplemented` (i.e. it's like `!isFunctionImported`). Maybe one or the other of these should be switched to make them consistent.

dschuff commented on src/s2wasm.h:
```diff
@@ -467,6 +470,7 @@ class S2WasmBuilder {
       else if (match("ident")) skipToEOL();
       else if (match("section")) parseToplevelSection();
       else if (match("align") || match("p2align")) skipToEOL();
+      else if (match("import_global")) parseImportGlobal();
```
since `parseImportGlobal` doesn't do anything, could this just be `skipToEOL()`?

jgravelle-google replied:
Yes, which more succinctly expresses the intent, and also fits nicely with the assembler directives on the previous line that we also just skip.

dschuff commented on src/wasm-emscripten.h:
```diff
@@ -29,6 +29,8 @@ void generateMemoryGrowthFunction(Module&);
 // signature in the indirect function table.
 std::vector<Function*> makeDynCallThunks(Module& wasm, std::vector<Name> const& tableSegmentData);
 
+void addGlobalImports(Module& wasm, std::unordered_set<Name> const& globalImports);
```
Maybe we should call these 'object imports' instead of 'global imports' since 'object' is the counterpart to 'function', the other kind of thing that can be imported (in the linker/ELF-like terminology that we are using). Functions and objects (i.e. the symbol's type in ELF terminology) can each be global or not (i.e. the symbol's binding in ELF terminology).

dschuff replied:
(And of course the fact that we use wasm globals to implement that is an implementation detail and also makes the term 'global' more confusing :))

dschuff commented on src/s2wasm.h:
```diff
@@ -834,6 +845,34 @@ class S2WasmBuilder {
       curr->finalize();
       setOutput(curr, assign);
     };
+    auto fixRelocation = [&](LinkerObject::Relocation* relocation,
```
Somewhere we should actually describe how imported globals work. Also it's a little weird that this function called `fixRelocation` doesn't really do anything to the relocation. Instead it either adds it unmodified, or generates some code (at which point it will never get added?). `addRelocation` might be marginally better; it seems like we could do even better than that, but it's not immediately obvious how).
Since we already handle null relocations, we could separate the `addRelocation` bit and put the interesting logic in some function that either returns the relocation unchanged, or generates code and returns nullptr? Not sure.

jgravelle-google replied:
`potentiallyFixUpExpressionIfRelocationWasActuallyDoneAndIsNotImportedOtherwiseJustAddRelocation`?

With the way I've written it, there's three cases. Let's use `makeLoad` as an example, creating a `Load` expression called `curr`:
1.  * `i32.load $push1=, 4($pop0)`
    * Parse up to the `,`
    * Call `getRelocatableConst` and get a numeric literal `4`
    * Set `curr->offset.addr` to `4`
    * Parse the `$pop0`, set `curr->ptr` to that expression
2.  * `i32.load $push1=, foo($pop0)`
    * Parse up to the `,`
    * Call `getRelocatableConst` and get a non-imported global `foo`
    * Create a relocation record for `foo`
    * Add relocation record to linker, to set `curr->offset.addr` to the global value at link-time
    * Parse the `$pop0`, set `curr->ptr` to that expression
3.  * `i32.load $push1=, external_foo($pop0)`
    * Parse up to the `,`
    * Call `getRelocatableConst` and get an imported global `external_foo`
    * Create a relocation record for `external_foo`
    * Parse the `$pop0`, set `curr->ptr` to that expression
    * Modify `curr->ptr` to add `get_global external_foo` and the `$pop0` expression [1]

So, due to the interleaving of parsing, stashing relocations for later, and constructing the expression, this is the factoring I came up with.
... *Having written all that out*, it occurs to me that we don't need to be passing around relocations here, and can instead have the first part return the expression that gets generated if we have an imported global, which reduces the jankiness somewhat. I'll push a change with that and we'll see how that looks.

[1]: Rereading the code that does this, it doesn't currently incorporate any `$pop0` expressions other than `i32.const 0`. I guess a test case is in order there.

jgravelle-google commented:
Added tests, addressed all comments.
Refactored to generate expressions more directly.

Added some minor optimizations to turn
```
i32.const $push0=, 0
i32.store foo+2($pop0), 7
```
from
```
(i32.store
  (i32.add
    (i32.const 0)
    (i32.add
      (i32.const 2)
      (get_global $foo)
    )
  )
  (i32.const 7)
)
```
into
```
(i32.store offset=2
  (get_global $foo)
  (i32.const 7)
)
```

dschuff commented on src/s2wasm.h:
```diff
@@ -229,29 +229,68 @@ class S2WasmBuilder {
+      s++;
+      offset = -getInt();
+    }
+    return new LinkerObject::Relocation(
```
ASan build says this is leaking.

jgravelle-google commented on src/s2wasm.h:
```diff
@@ -229,29 +229,68 @@ class S2WasmBuilder {
+    // Only do this for positive addends because load/store offsets cannot be
+    // negative.
+    if (relocation->addend >= 0) {
+      *relocation->data = relocation->addend;
```
relocation->data is the `target` pointer that we pass in to `getRelocatableConst`. Given that it's a field on an expression that we're in the middle of creating, it should be either 0 or junk.
For a non-imported global, we clobber that value when we do the relocation in wasm-linker.cpp, so this is morally equivalent.

jgravelle-google replied:
Trivia: there's an interesting, annoyingly-subtle interaction here.
There's four places that call `getRelocatableConst`: load, store, const, and data section initializer.
In data, we don't model `get_global` at all, so we abort here for now.
For load and store, this writes to the `offset=___` field, and is why we do this.
Const is the odd one, because the pointer we write to is the value of the const we're creating. Without this that would become an `(i32.const 0)`, that the other optimization would remove.
Either way we turn `i32.const $push0=, foo+2` into `(i32.add (i32.const 2) (get_global $foo))`

I imagine this is quirky enough that it will someday become a problem, and it's probably worth just emitting a bunch of `(i32.add (i32.const 0)`s and letting the optimizer take care of it, just for sanity's sake.

dschuff replied:
I'm not sure LLVM can generate `i32.const foo+2` anyway. I think this is fine for now because we will end up revisiting it for MC and linking support anyway.

sunfishcode replied:
FWIW, LLVM does generate that, from code like this:
```
extern int foo[];

int* foo(void) {
    return &foo[2];
}
```

jgravelle-google replied:
Or from code like:

```c
struct Foo {
    int a;
    int b;
    int c;
};
extern struct Foo foo;

int main() {
  printf("foo: %p\n", &foo.c);
  return 0;
}
```

Having thought about this more, I believe each piece of it works in isolation. We write to the constant expression we generate. We preserve the existing expression when we add the `get_global`. This isn't actually *unsafe* in that case, just oddly redundant. Going to keep it for now.

kripken commented:
> Added some minor optimizations

Those seem like things the optimizer should do, but looks like it didn't yet, so I wrote #850. With that, you could rely on the optimizer to do it for you if you want.
